File: 7_last_digit_of_the_sum_of_fibonacci_numbers_again/fibonacci_partial_sum.py
"""
Author: Jehadur Rahman
Organization: CyArm
GitHub: https://github.com/JehadurRE
Website: jehadurre.me
Date: 2025-07-10 09:43 (Thu)
"""

# Uses python3
import sys

def fibonacci_partial_sum_naive(from_, to):
    _sum = 0

    current = 0
    _next = 1

    for i in range(to + 1):
        if i >= from_:
            _sum += current

        current, _next = _next, current + _next

    return _sum % 10

def fibonacci_partial_sum_fast(from_, to):
    sum_val = 0
    
    current = 0
    next_val = 1
    from_ = from_ % 60
    to = to % 60
    
    if to < from_:
        to += 60
    
    for i in range(to + 1):
        if i >= from_:
            sum_val += current
        
        new_current = next_val
        next_val = (next_val + current) % 10
        current = new_current
    
    return sum_val % 10

if __name__ == '__main__':
    input_data = sys.stdin.read()
    from_, to = map(int, input_data.split())
    print(fibonacci_partial_sum_fast(from_, to))

File: 7_last_digit_of_the_sum_of_fibonacci_numbers_again/fibonacci_partial_sum.hs
-- Author: Jehadur Rahman
-- Organization: CyArm
-- GitHub: https://github.com/JehadurRE
-- Website: jehadurre.me
-- Date: 2025-07-10 09:43 (Thu)

-- by Kirill Elagin

fibonacci_partial_sum_naive :: Integer -> Integer -> Int
fibonacci_partial_sum_naive from to = let (a', b', _) = helper (0, 1, 0) from
                                          (_, _, s) = helper (a', b', a') (to - from)
                                      in s
  where
    helper (a, b, s) 0 = (a, b, s `mod` 10)
    helper (a, b, s) i = helper (b, a + b, s + b) (i - 1)

fibonacci_partial_sum_fast :: Integer -> Integer -> Int
fibonacci_partial_sum_fast from to = helper (0, 1, 0) 0
  where
    from' = from `mod` 60
    to' = if (to `mod` 60) < from' then (to `mod` 60) + 60 else to `mod` 60
    helper (current, next, sum) i
      | i > to' = sum `mod` 10
      | i >= from' = helper (next, (current + next) `mod` 10, (sum + current) `mod` 10) (i + 1)
      | otherwise = helper (next, (current + next) `mod` 10, sum) (i + 1)

main :: IO ()
main = do
  [from, to] <- fmap words getLine
  print $ fibonacci_partial_sum_fast (read from) (read to)
